### PR TITLE
Feature mtc 297 temporary capital

### PR DESCRIPTION
**Pull request by GSailen**

### CODE REVIEW

File: src/MarginTrading.AccountsManagement.Contracts/ITemporaryCapitalController.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+using MarginTrading.AccountsManagement.Contracts.Api;
+using MarginTrading.AccountsManagement.Contracts.Models;
+using Refit;
+
+namespace MarginTrading.AccountsManagement.Contracts
+{
+    /// <summary>
+    /// Manage user temporary capital
+    /// </summary>
+    public interface ITemporaryCapitalController
```
Comment (Waqas-Zamir) [inline src/MarginTrading.AccountsManagement.Contracts/ITemporaryCapitalController.cs]: Some endpoint to approve temporary capital request is also required.
But I guess that part could be done on Lego side probably in donut service.

Comment (GSailen) [reply src/MarginTrading.AccountsManagement.Contracts/ITemporaryCapitalController.cs]: That's correct

File: src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalCommandsHandler.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                    nameof(StartRevokeTemporaryCapitalInternalCommand), exception);
+                
+                publisher.PublishEvent(new RevokeTemporaryCapitalFailedEvent(c.OperationId, 
+                    _systemClock.UtcNow.UtcDateTime, exception.Message, c.EventSourceId, 
+                    c.RevokeEventSourceId));
+                
+                return;
+            }
+
+            publisher.PublishEvent(new RevokeTemporaryCapitalStartedInternalEvent(c.OperationId,
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalCommandsHandler.cs]: Kitty is missed

File: src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalCommandsHandler.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                    revokeEventSourceId: executionInfo.Data.RevokeEventSourceId,
+                    auditLog: executionInfo.Data.AuditLog));
+            }
+            else if (executionInfo.Data.State == TemporaryCapitalState.Failing)
+            {
+                //rollback account. if exception is thrown here, it will be retried until success
+                await _accountsRepository.UpdateAccountRollbackTemporaryCapitalAsync(executionInfo.Data.AccountId,
+                    executionInfo.Data.RevokedTemporaryCapital);
+                
+                publisher.PublishEvent(new RevokeTemporaryCapitalFailedEvent(c.OperationId,
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalCommandsHandler.cs]: Kitty again

File: src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalSaga.cs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            if (executionInfo.Data.SwitchState(TemporaryCapitalState.Initiated, TemporaryCapitalState.Started))
+            {
+                executionInfo.Data.RevokedTemporaryCapital = e.RevokedTemporaryCapital;
+                
+                sender.SendCommand(
+                    new UpdateBalanceInternalCommand(
+                        operationId: e.OperationId,
+                        accountId: executionInfo.Data.AccountId,
+                        amountDelta: - executionInfo.Data.RevokedTemporaryCapital.Sum(x => x.Amount),
+                        comment: "Revoke temporary capital",
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalSaga.cs]: Let's put revoked operations somewhere... Maybe into the comment field.

File: src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalSaga.cs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+        /// <summary>
+        /// Operation failed
+        /// </summary>
+        [UsedImplicitly]
+        private async Task Handle(RevokeTemporaryCapitalFailedEvent e, ICommandSender sender)
+        {
+            var executionInfo = await _executionInfoRepository.GetAsync<RevokeTemporaryCapitalData>(OperationName, e.OperationId);
+
+            if (executionInfo != null && executionInfo.Data.SwitchState(TemporaryCapitalState.Initiated,
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalSaga.cs]: Failing state is missing

File: src/MarginTrading.AccountsManagement.Contracts/ITemporaryCapitalController.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        /// </summary>
+        /// <param name="request"></param>
+        [Delete("/api/temporary-capital")]
+        Task<string> RevokeTemporaryCapital([Body] RevokeTemporaryCapitalRequest request);
+
+        /// <summary>
+        /// See the list of transactions of temporary capital
+        /// </summary>
+        [Get("/api/temporary-capital/{accountId}")]
+        Task<List<AccountBalanceChangeContract>> ListTemporaryCapital(string accountId,
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AccountsManagement.Contracts/ITemporaryCapitalController.cs]: +

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs [hunk 7]
```diff
@@ -63,16 +69,23 @@ public AccountBalanceChangesRepository(IConvertService convertService, AccountMa
...
         
         public async Task<IReadOnlyList<IAccountBalanceChange>> GetAsync(string accountId, DateTime? @from = null,
-            DateTime? to = null)
+            DateTime? to = null, AccountBalanceChangeReasonType? reasonType = null)
         {
             var whereClause = "WHERE 1=1 " + (!string.IsNullOrWhiteSpace(accountId) ? " AND AccountId=@accountId" : "")
                                        + (from != null ? " AND ChangeTimestamp > @from" : "")
-                                       + (to != null ? " AND ChangeTimestamp < @to" : "");
+                                       + (to != null ? " AND ChangeTimestamp < @to" : "")
+                                       + (reasonType != null ? " AND ReasonType = @reasonType" : "");
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs]: Good idea

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs [hunk 8]
```diff
@@ -93,6 +106,16 @@ public async Task<IReadOnlyList<IAccountBalanceChange>> GetAsync(string accountI
...
         }
 
+        public async Task<decimal> GetRealizedDailyPnl(string accountId)
+        {
+            var history = await GetAsync(
+                accountId: accountId,
+                //TODO rethink the way trading day's start & end are selected 
+                @from: _systemClock.UtcNow.UtcDateTime.Date,
+                reasonType: AccountBalanceChangeReasonType.RealizedPnL);
+            return history.Sum(x => x.ChangeAmount);
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountBalanceChangesRepository.cs]: It's faster to calc everything on side of SQL

File: src/MarginTrading.AccountsManagement/Repositories/IAccountsRepository.cs [hunk 9]
```diff
@@ -27,7 +27,12 @@ internal interface IAccountsRepository
         Task<IAccount> UpdateBalanceAsync(string operationId, string accountId,
             decimal amountDelta, bool changeLimit);
 
-        Task<IAccount> UpdateAccountAsync(string accountId,
-            string tradingConditionId, bool? isDisabled, bool? isWithdrawalDisabled);
+        Task<IAccount> UpdateAccountAsync(string accountId, string tradingConditionId,
+            bool? isDisabled, bool? isWithdrawalDisabled);
+
+        Task<IAccount> UpdateAccountTemporaryCapitalAsync(string accountId, TemporaryCapital temporaryCapital, 
+            bool addOrRemove);
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/IAccountsRepository.cs]: in that case maybe isAddOrRemove ? )

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs [hunk 10]
```diff
@@ -171,7 +172,57 @@ public async Task<IAccount> GetAsync(string accountId)
...
-        
+
+        public async Task<IAccount> UpdateAccountTemporaryCapitalAsync(string accountId,
+            TemporaryCapital temporaryCapital, bool addOrRemove)
+        {
+            return await GetAccountAndUpdate(accountId, a =>
+            {
+                var result = ((IAccount) a).TemporaryCapital;
+
+                if (addOrRemove)
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/AccountsRepository.cs]: ok lets pass as Func from Service

File: src/MarginTrading.AccountsManagement/Workflow/GiveTemporaryCapital/GiveTemporaryCapitalCommandsHandler.cs [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            var account = await _accountsRepository.GetAsync(c.AccountId);
+
+            if (account == null)
+            {
+                publisher.PublishEvent(new GiveTemporaryCapitalFailedEvent(c.OperationId, 
+                    _systemClock.UtcNow.UtcDateTime, $"Account {c.AccountId} not found", c.EventSourceId));
+                return;
+            }
+
+            if (account.IsDisabled)
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Workflow/GiveTemporaryCapital/GiveTemporaryCapitalCommandsHandler.cs]: Ok

File: src/MarginTrading.AccountsManagement/Workflow/GiveTemporaryCapital/GiveTemporaryCapitalCommandsHandler.cs [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+                await _log.WriteErrorAsync(nameof(GiveTemporaryCapitalCommandsHandler),
+                    nameof(StartGiveTemporaryCapitalInternalCommand), exception);
+                
+                publisher.PublishEvent(new GiveTemporaryCapitalFailedEvent(c.OperationId, 
+                    _systemClock.UtcNow.UtcDateTime, exception.Message, c.EventSourceId));
+                
+                return;
+            }
+
+            publisher.PublishEvent(new GiveTemporaryCapitalStartedInternalEvent(c.OperationId,
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Workflow/GiveTemporaryCapital/GiveTemporaryCapitalCommandsHandler.cs]: Ok

File: src/MarginTrading.AccountsManagement/Repositories/IAccountsRepository.cs [hunk 13]
```diff
@@ -27,7 +27,12 @@ internal interface IAccountsRepository
...
+
+        Task<IAccount> UpdateAccountRollbackTemporaryCapitalAsync(string accountId, List<TemporaryCapital> revokedTemporaryCapital);
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Repositories/IAccountsRepository.cs]: Ok

File: src/MarginTrading.AccountsManagement/Workflow/GiveTemporaryCapital/GiveTemporaryCapitalCommandsHandler.cs [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        /// </summary>
+        [UsedImplicitly]
+        public async Task Handle(FinishGiveTemporaryCapitalInternalCommand c, IEventPublisher publisher)
+        {
+            var executionInfo = await _executionInfoRepository.GetAsync<GiveTemporaryCapitalData>(OperationName, c.OperationId);
+
+            if (executionInfo == null)
+                return;
+            
+            _chaosKitty.Meow(c.OperationId);
```
[the remark this answers is not in the document]
Comment (GSailen) [reply src/MarginTrading.AccountsManagement/Workflow/GiveTemporaryCapital/GiveTemporaryCapitalCommandsHandler.cs]: Ok

File: src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs [hunk 15]
```diff
@@ -113,7 +113,9 @@ internal class OperationExecutionInfoRepository : IOperationExecutionInfoReposit
                 catch (SqlException)
                 {
                     await conn.ExecuteAsync(
-                        $"update {TableName} set {GetUpdateClause} where Id=@Id and OperationName=@OperationName", entity);
+                        $"update {TableName} set {GetUpdateClause} where Id=@Id " +
```
Comment (andrey-anokhin) [inline src/MarginTrading.AccountsManagement/Repositories/Implementation/SQL/OperationExecutionInfoRepository.cs]: Please copy the whole method Save from Commissions Service or Trading Core ;)

File: src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalCommandsHandler.cs (hunk 3 above)
Comment (andrey-anokhin) [reply src/MarginTrading.AccountsManagement/Workflow/RevokeTemporaryCapital/RevokeTemporaryCapitalCommandsHandler.cs]: Not fixed

**Pull request merged**